### DISCUSSION TITLE
Looking for more straightforward way to extract svg

### DISCUSSION THREAD

**Discussion by Detsudetsu** [pymupdf/PyMuPDF#1175] [Looking for help]
I'm trying to extract drawings from a PDF file and save them as SVG.
Thanks to a previous work found in [here](https://github.com/pymupdf/PyMuPDF/issues/178#issuecomment-872160813), I managed, but the implementation feels a bit roundabout to me.
What my current implementation below does is extract drawings, get their rect size, crop the source pdf to that size, and finally `get_svg_image`; such a long way!

I want to find any better way than that. Any ideas?

```py
import fitz
import sys
from pathlib import Path

doc = fitz.open(sys.argv[1])
path = Path(sys.argv[1])
dst_dir = Path(path).with_suffix("").with_stem(path.stem + "-drawings")
dst_dir.mkdir()
out = fitz.open()
for page in doc:
    new_rects = []  # resulting rectangles

    for p in page.get_drawings():
        w = p["width"]
        r = p["rect"] + (-0 * w, -0 * w, 0 * w, 0 * w)  # enlarge each rectangle by width value
        for i in range(len(new_rects)):
            if abs(r & new_rects[i]) > 0:  # touching one of the new rects?
                new_rects[i] |= r  # enlarge it
                break
        # now look if contained in one of the new rects
        remainder = [s for s in new_rects if r in s]
        if remainder == []:  # no ==> add this rect to new rects
            new_rects.append(r)

    new_rects = list(set(new_rects))  # remove any duplicates
    new_rects.sort(key=lambda r: abs(r), reverse=True)
    remove = []
    for j in range(len(new_rects)):
        for i in range(len(new_rects)):
            if new_rects[j] in new_rects[i] and i != j:
                remove.append(j)
    remove = list(set(remove))
    remove.sort(reverse=True)
    for i in remove:
        del new_rects[i]
    new_rects.sort(key=lambda r: (r.tl.y, r.tl.x))  # sort by location

    for i, r in enumerate(new_rects):
        if r.height <= 5 or r.width <= 5:
            continue  # skip lines and empty rects

        newpage = out.new_page(-1, r.width, r.height)
        newpage.show_pdf_page(page.rect, doc, page.number, clip=r)
        svg = newpage.get_svg_image()
        file_name = "drawing-%03i-%02i.svg" % (page.number, i)
        dst_path = dst_dir / file_name
        dst_path.write_text(svg)

        print("saved: ", dst_path)
```

**Comment by JorjMcKie** [comment]
I actually like your code 😎
The most complex part is identifying the relevant drawing rectangles, removing containments, duplicates and all that.
I see no way how to simplify this.

But you probably need no dummy `out` PDF and those `newpages`:
Why not setting the CropBox of the original page to the identified rectangle and directly make an SVG from the original page?
```python
    ... 
    page.set_cropbox(r)
    svg = page.get_svg_image()
    ...
```

**Reply by Detsudetsu** [reply]
I'd like to accept your suggestion, which makes the code much clearer!
But let me share a minor issue I found.

I found that some applications wrongly display even outside of the identified rectangle of the SVG created via the `CropBox` method, but not via the `newpage` one.
That's because what the former method does is just set `viewPort` and `viewBox`, so it doesn't do any clipping actually.
On the other hand, the latter method creates a `<clipPath>` element corresponding to the rectangle besides what the former does.

For example, the following screenshot shows SVG preview presented by [PowerToys File Explorer Add-ons](https://aka.ms/PowerToysOverview_FileExplorerAddOns).
We can see the previews of rectangles (not a square) include irrelevant stuff in the upper directory, containing SVG created via the `CropBox` method.

In that cases, I have to **clip** the page, not **crop**.
So `show_pdf_page()` is still an option, perhaps?

![Screenshot](https://user-images.githubusercontent.com/36166146/127761950-a4d04378-51c2-4a77-ad8c-f89b579346cc.png)

[sample.pdf](https://github.com/pymupdf/PyMuPDF/files/6912246/sample.pdf)

**Reply by JorjMcKie** [reply]
Ah, ok, interesting. Didn't know you tried that already. You must be a PyMuPDF power user 😉.
Yes, the `show_pdf_page()` is the only option to "set" the MediaBox of a page.
I tried setting TrimBox, ArtBox and BleedBox to the value of CropBox after updating the latter (via `doc.xref_set_key()`), but the SVG sources contain a lot of unexpected (text) stuff in any case - don't know if they would show up under PowerToys (haven't installed that one) when all the other PDF boxes also have been set ...

**Reply by Detsudetsu** [reply]
Alright now I tried setting TrimBox, ArtBox and BleedBox and yet it didn't work so might as well go with `show_pdf_page()`.
I really appreciate your help!

**Comment by Detsudetsu** [comment]
By the way, I did a little experiment to find which bounding box (like MediaBox) can clip as I wanted.
I added the following snippet to the very first code.

```py
for page in out:
    xref = out.page_xref(page.number)
    print("page:", page.number)
    for key in doc.xref_get_keys(xref):
        print(key, "=" , doc.xref_get_key(xref, key))
```

Then ran the code with the sample.pdf that was attached in my previous comment (containing 3 blue quadrangles)
The result was below.

```
page: 0
Filter = ('name', '/FlateDecode')
Length = ('int', '3106')
page: 1
Type = ('name', '/ObjStm')
N = ('int', '24')
First = ('int', '172')
Filter = ('name', '/FlateDecode')
Length = ('int', '551')
page: 2
P = ('xref', '24 0 R')
S = ('name', '/P')
Type = ('name', '/StructElem')
K = ('array', '[27 0 R 28 0 R]')
Pg = ('xref', '3 0 R')
```

It's chaotic 🤯
How can they be so different?

**Comment by JorjMcKie** [comment]
> It's chaotic 🤯
How can they be so different?

Don't know what you have been doing ... 🤔
When I use this script, all works fine and looks decent:
```python
import fitz

doc = fitz.open("sample.pdf")
page = doc[0]
paths = page.get_drawings()
src = fitz.open()
for i, path in enumerate(paths):
    r = path["rect"]
    spage = src.new_page(width=r.width, height=r.height)
    spage.show_pdf_page(spage.rect, doc, page.number, clip=r)
    svg = spage.get_svg_image(text_as_path=False)
    out = open("%i-new.svg" % i, "w")
    out.write(svg)
    out.close()

for page in src:
    keys = src.xref_get_keys(page.xref)
    print("\nTempfile page", page.number, "\n")
    for k in keys:
        print(k, src.xref_get_key(page.xref, k))
```
Result:
```
Tempfile page 0 

Type ('name', '/Page')
MediaBox ('array', '[0 0 147.4 147.40003]')
Rotate ('int', '0')
Resources ('xref', '3 0 R')
Parent ('xref', '2 0 R')
Contents ('array', '[20 0 R]')

Tempfile page 1 

Type ('name', '/Page')
MediaBox ('array', '[0 0 147.4 73.70001]')
Rotate ('int', '0')
Resources ('xref', '21 0 R')
Parent ('xref', '2 0 R')
Contents ('array', '[24 0 R]')

Tempfile page 2 

Type ('name', '/Page')
MediaBox ('array', '[0 0 73.69999 147.39997]')
Rotate ('int', '0')
Resources ('xref', '25 0 R')
Parent ('xref', '2 0 R')
Contents ('array', '[28 0 R]')
```

**Reply by Detsudetsu** [reply]
Oops I've found it 😑
I should have used `out.xref_get_key()` instead of `doc.xref_get_key()`.
And I've got the same result as yours.

Thank you for your help, and sorry for the late reply.
